Make ec_ arithmetic more consistent and add documentation

jonasnick opened the pull request:
Fixes #671. Supersedes #668.

This PR unifies handling of invalid secret keys by introducing a new function `scalar_set_b32_secret` which returns false if the b32 overflows or is 0. By using this in `privkey_{negate, tweak_add, tweak_mul}` these function will now return 0 if the secret key is invalid which matches the behavior of `ecdsa_sign` and `pubkey_create`.

Instead of deciding whether to zeroize the secret key on failure, I only added documentation for now that the value is undefined on failure.

real-or-random commented on include/secp256k1.h:
```diff
@@ -599,9 +602,10 @@ SECP256K1_API SECP256K1_WARN_UNUSED_RESULT int secp256k1_ec_pubkey_negate(
- *          private key). 1 otherwise.
+ *          uniformly random 32-byte arrays, or if the given private key is
+ *          invalid according to secp256k1_ec_seckey_verify, or if the resulting
+ *          private key would be invalid (only when the tweak is the complement
```
Maybe "negated private key" instead of "complement"?

Is this really the only case? What about overflows?

real-or-random replied:
This should also document what the valid range for a tweak is? (I guess every scalar?)

apoelstra replied:
Also, missing an `)` after "32-byte arrays". (This was a preexisting problem but we should fix it now.)

jonasnick replied:
@apoelstra should be fixed in the latest commit.

real-or-random commented on include/secp256k1.h:
```diff
@@ -614,9 +618,10 @@ SECP256K1_API SECP256K1_WARN_UNUSED_RESULT int secp256k1_ec_privkey_tweak_add(
- *          corresponding private key). 1 otherwise.
+ *          uniformly random 32-byte arrays, or if the given private key is
+ *          invalid according to secp256k1_ec_seckey_verify, or if the resulting
+ *          public key would be invalid (only when the tweak is the complement
```
same here

apoelstra replied:
same `)` issue here

real-or-random commented on src/secp256k1.c:
```diff
@@ -526,7 +526,9 @@ int secp256k1_ec_privkey_negate(const secp256k1_context* ctx, unsigned char *sec
     ARG_CHECK(seckey != NULL);
 
-    secp256k1_scalar_set_b32(&sec, seckey, NULL);
+    if(!secp256k1_scalar_set_b32_seckey(&sec, seckey)) {
```
nit: spacing

jonasnick replied:
fixed

real-or-random commented on include/secp256k1.h:
```diff
@@ -584,7 +584,8 @@ SECP256K1_API SECP256K1_WARN_UNUSED_RESULT int secp256k1_ec_pubkey_create(
  *  In/Out: seckey:     pointer to the 32-byte private key to be negated. The private
- *                      key should be valid according to secp256k1_ec_seckey_verify
+ *                      key should be valid according to secp256k1_ec_seckey_verify.
+ *                      Value becomes undefined if this function returns 0.
```
Can we say "unspecified"? "Undefined" comes with that weird terrifying sound of undefined behavior.

jonasnick commented:
@real-or-random  I added a commit that rephrases the docs. Let me know if you think that's better.

(Also, I replaced `secp256k1_rand256_test` with `secp256k1_rand256` because the former is apparently not random, may produce invalid seckeys and travis randomly hit that in one of the test configurations).

jonasnick commented on src/tests.c:
```diff
@@ -1117,16 +1117,43 @@ void run_scalar_tests(void) {
+        secp256k1_scalar_set_b32(&scalar, bin, &overflow);
+        CHECK(overflow == 0);
+        secp256k1_scalar_get_b32(bin_tmp, &scalar);
+        CHECK(memcmp(bin, bin_tmp, 32) == 0);
```
It's probably the only roundtrip because it's the only one which doesn't overlow and which can not be easily tested with `scalar_is_zero`.

jonasnick commented:
Removed the (attempted) constant time AND in `set_b32_seckey` as per @sipa's suggestion

real-or-random commented:
Is there a reason to believe that this will break downstream code (because it changes the API)?

dr-orlovsky commented on include/secp256k1.h:
```diff
@@ -597,57 +611,85 @@ SECP256K1_API SECP256K1_WARN_UNUSED_RESULT int secp256k1_ec_pubkey_negate(
+ * In:     tweak:  pointer to a 32-byte tweak. Must be in the same range as secret
+ *                 keys (see secp256k1_ec_seckey_verify). For uniformly random
+ *                 32-byte arrays the chance of being out of range is
+ *                 negligible (around 1 in 2^128). (cannot be NULL)
```
while the changes are negligible, if one is aimed to write a deterministic code (like in the case of rust wrapper), it may be good to cover this case as well. 

It is possible to add to the docs that in case `tweak`<Z_p, the function will return 0?
Also, do you see any value of differentiating between `tweak`<Z_p and point at infinity cases? Right now both will return 0, but from the end-user perspective the first case requires change of the tweaking factor, while the second may be solved just by using a different public key...

real-or-random replied:
> while the changes are negligible, if one is aimed to write a deterministic code (like in the case of rust wrapper), it may be good to cover this case as well.

Believe me, you don't care about a failure probability of 2^-128. It's astronomically small. It's much smaller than the probability of a hardware error (e.g., RAM returns wrong value). You can simply abort in this case.

apoelstra commented on src/tests.c:
```diff
@@ -3994,6 +3994,16 @@ void run_eckey_edge_case_test(void) {
+    memset(ctmp2, 0, 32);
+    ctmp2[31] = 0x01;
+    CHECK(secp256k1_ec_seckey_verify(ctx, ctmp2) == 1);
+    CHECK(secp256k1_ec_privkey_tweak_add(ctx, ctmp, ctmp2) == 0);
```
Can you add a check with `ctmp` and `ctmp2` swapped? Or maybe there is little value in this since both of these are invalid.

jonasnick replied:
Are you suggesting to test this function with an overflowing tweak? That should happen a couple of lines below. I pushed a commit to make the comments more clear.

gmaxwell commented:
@jonasnick  the _test rng's are biased intentionally to hit corner cases and it worked perfectly there if it detected that your scalar generation was generating an invalid scalar.  Maybe you want random_scalar_order_test  or random_scalar_order instead?   It's better to make the tests not fail even with rare values because they can also be moved over to smaller groups where the failures would be common.

jonasnick commented:
@gmaxwell Thanks, `random_scalar_order` is almost what I was looking for and agree that it's better than `rand256`. I added a function similar to `random_scalar_order` but returning a byte array.

real-or-random commented:
Can you squash this?

jonasnick commented:
Rebasing required a bit of thought due to #710. Please have a look.

jonasnick commented:
I improved `ecdsa_sign`'s uses of `secp256k1_scalar_set_b32_seckey` a little bit.

real-or-random commented on src/tests.c:
```diff
@@ -3936,45 +3989,64 @@ void run_eckey_edge_case_test(void) {
     memcpy(&pubkey, &pubkey2, sizeof(pubkey));
-    /* Overflowing key tweak zeroizes. */
+    /* If seckey_tweak_add or seckey_tweak_mul are called with an overflowing
+    seckey, the seckey is zeroized. */
```
Here they're not called with an overflowing seckey but the resulting seckey would overflow.

jonasnick replied:
No, seckey is equal to orderc which is overflowing. Added `CHECK(secp256k1_ec_seckey_verify(ctx, ctmp) == 0);`

real-or-random commented on src/scalar_impl.h:
```diff
@@ -55,6 +55,12 @@ static void secp256k1_scalar_order_get_num(secp256k1_num *r) {
+static int secp256k1_scalar_set_b32_seckey(secp256k1_scalar *r, const unsigned char *bin) {
+  int overflow;
+  secp256k1_scalar_set_b32(r, bin, &overflow);
+  return (!overflow) & (!secp256k1_scalar_is_zero(r));
```
nit: indentation
I only point it out here but it's wrong in many functions you added.

jonasnick replied:
Ugh, fixed, but only saw this in one other function (run_scalar_set_b32_seckey_tests)

real-or-random commented on include/secp256k1.h:
```diff
@@ -569,20 +572,31 @@ SECP256K1_API SECP256K1_WARN_UNUSED_RESULT int secp256k1_ec_seckey_verify(
+) SECP256K1_ARG_NONNULL(1) SECP256K1_ARG_NONNULL(2);
+
+/** Same as secp256k1_ec_seckey_negate, but DEPRECATED. Will be removed in
+ *  future versions. */
```
Can you change the calls in the codebase to  the `seckey` variants? For example https://github.com/bitcoin-core/secp256k1/blob/master/src/valgrind_ctime_test.c#L76

jonasnick replied:
Done

real-or-random commented on include/secp256k1.h:
```diff
@@ -599,57 +613,85 @@ SECP256K1_API SECP256K1_WARN_UNUSED_RESULT int secp256k1_ec_pubkey_negate(
- *          uniformly random 32-byte arrays, or if the resulting private key
- *          would be invalid (only when the tweak is the complement of the
- *          private key). 1 otherwise.
+/** Tweak a secret key by adding tweak to it.
```
nit: Now that you're rewriting those, can you add an empty line here? (same for the other tweak functions)

jonasnick replied:
Done

real-or-random commented on src/eckey_impl.h:
```diff
@@ -54,10 +54,7 @@ static int secp256k1_eckey_pubkey_serialize(secp256k1_ge *elem, unsigned char *p
-    if (secp256k1_scalar_is_zero(key)) {
-        return 0;
-    }
-    return 1;
```
Huh, shouldn't this have been caught by the new constant time test?

jonasnick replied:
Presumably the compiler optimized away the branching

real-or-random commented on include/secp256k1.h:
```diff
@@ -599,57 +613,85 @@ SECP256K1_API SECP256K1_WARN_UNUSED_RESULT int secp256k1_ec_pubkey_negate(
- *          would be invalid (only when the tweak is the complement of the
- *          corresponding private key). 1 otherwise.
+ * Returns: 0 if the resulting public key would be invalid (only when the tweak
+ *          is the negation of the corresponding secret key). 1 otherwise.
```
What if the tweak itself is out of range?

jonasnick replied:
Added a commit to clarify

real-or-random commented on include/secp256k1.h:
```diff
@@ -569,20 +572,31 @@ SECP256K1_API SECP256K1_WARN_UNUSED_RESULT int secp256k1_ec_seckey_verify(
  *  Args:   ctx:        pointer to a context object
- *  In/Out: seckey:     pointer to the 32-byte private key to be negated (cannot be NULL)
+ *  In/Out: seckey:     pointer to the 32-byte secret key to be negated. The secret
+ *                      key should be valid according to secp256k1_ec_seckey_verify.
```
nit: "should" is somewhat imprecise. suggestion: "If the secret key is invalid according to secp256k1_ec_seckey_verify, this function returns 0 and seckey will be set to some  unspecified value."

jonasnick replied:
fixed in separate commit

real-or-random commented on include/secp256k1.h:
```diff
@@ -599,57 +613,85 @@ SECP256K1_API SECP256K1_WARN_UNUSED_RESULT int secp256k1_ec_pubkey_negate(
- * In/Out:  seckey: pointer to a 32-byte private key.
- * In:      tweak:  pointer to a 32-byte tweak.
- */
+ * In/Out:  seckey: pointer to a 32-byte secret key. The secret key should be
```
same as above

jonasnick replied:
fixed

real-or-random commented on include/secp256k1.h:
```diff
@@ -599,57 +613,85 @@ SECP256K1_API SECP256K1_WARN_UNUSED_RESULT int secp256k1_ec_pubkey_negate(
- *          would be invalid (only when the tweak is the complement of the
- *          private key). 1 otherwise.
+/** Tweak a secret key by adding tweak to it.
+ * Returns: 0 if the resulting secret key would be invalid (only when the tweak
```
What if the tweak itself is out of range?

jonasnick replied:
fixed

real-or-random commented on include/secp256k1.h:
```diff
@@ -599,57 +613,85 @@ SECP256K1_API SECP256K1_WARN_UNUSED_RESULT int secp256k1_ec_pubkey_negate(
+ *                  valid according to secp256k1_ec_seckey_verify. If this
+ *                  function returns 0, seckey will be some unspecified
+ *                  value. (cannot be NULL)
+ * In:       tweak: pointer to a 32-byte tweak. Must be in the same range as secret
```
"must" is similar to "should" here. I guess it's better to specify the consequences if it's not in range, then the function will return 0.

jonasnick replied:
fixed

jonasnick commented on include/secp256k1.h:
```diff
@@ -551,7 +551,10 @@ SECP256K1_API int secp256k1_ecdsa_sign(
 ) SECP256K1_ARG_NONNULL(1) SECP256K1_ARG_NONNULL(2) SECP256K1_ARG_NONNULL(3) SECP256K1_ARG_NONNULL(4);
 
-/** Verify an ECDSA secret key.
+/** Verify an ECDSA secret key. It is valid if it is not 0 and less than the
```
fixed

jonasnick commented on include/secp256k1.h:
```diff
@@ -551,7 +551,10 @@ SECP256K1_API int secp256k1_ecdsa_sign(
-/** Verify an ECDSA secret key.
+/** Verify an ECDSA secret key. It is valid if it is not 0 and less than the
+ *  secp256k1 curve order when interpreted as an integer (most significant byte
+ *  first). The probability of choosing a 32-byte string at random which is an
```
fixed

jonasnick commented on include/secp256k1.h:
```diff
@@ -599,57 +613,85 @@ SECP256K1_API SECP256K1_WARN_UNUSED_RESULT int secp256k1_ec_pubkey_negate(
+ * In:       tweak: pointer to a 32-byte tweak. Must be in the same range as secret
+ *                  keys (see secp256k1_ec_seckey_verify). For uniformly random
+ *                  32-byte arrays the chance of being out of range is
+ *                  negligible (around 1 in 2^128). (cannot be NULL)
```
I don't understand what you're suggesting. Yes, the tweak must be in range. The resulting seckey will be in range if the tweak is not the negation of the seckey.

jonasnick commented on src/secp256k1.c:
```diff
@@ -471,7 +471,7 @@ int secp256k1_ecdsa_sign(const secp256k1_context* ctx, secp256k1_ecdsa_signature
     secp256k1_scalar sec, non, msg;
     int ret = 0;
-    int overflow = 0;
+    int is_sec_valid;
```
Afaics it's just wasted work if we fix the duplicated recovery code. Better to make recovery use the common code in a separate PR (right after this one).

real-or-random commented on src/secp256k1.c:
```diff
@@ -556,17 +554,23 @@ int secp256k1_ec_pubkey_create(const secp256k1_context* ctx, secp256k1_pubkey *p
+}
+
+int secp256k1_ec_privkey_negate(const secp256k1_context* ctx, unsigned char *seckey) {
+  return secp256k1_ec_seckey_negate(ctx, seckey);
```
nit: indentation

jonasnick replied:
fixed

real-or-random commented on src/secp256k1.c:
```diff
@@ -605,6 +609,10 @@ int secp256k1_ec_privkey_tweak_add(const secp256k1_context* ctx, unsigned char *
 }
 
+int secp256k1_ec_privkey_tweak_add(const secp256k1_context* ctx, unsigned char *seckey, const unsigned char *tweak) {
+  return secp256k1_ec_seckey_tweak_add(ctx, seckey, tweak);
```
here too

jonasnick replied:
fixed

real-or-random commented on src/secp256k1.c:
```diff
@@ -649,6 +657,10 @@ int secp256k1_ec_privkey_tweak_mul(const secp256k1_context* ctx, unsigned char *
 }
 
+int secp256k1_ec_privkey_tweak_mul(const secp256k1_context* ctx, unsigned char *seckey, const unsigned char *tweak) {
+  return secp256k1_ec_seckey_tweak_mul(ctx, seckey, tweak);
```
same

jonasnick replied:
fixed

real-or-random commented on include/secp256k1.h:
```diff
@@ -599,57 +616,93 @@ SECP256K1_API SECP256K1_WARN_UNUSED_RESULT int secp256k1_ec_pubkey_negate(
- */
+/** Tweak a secret key by adding tweak to it.
+ *
+ *  Returns: 0 if the arguments are invalid or resulting secret key would be
```
```suggestion
 *  Returns: 0 if the arguments are invalid or the resulting secret key would be
```

jonasnick replied:
fixed

jonasnick commented:
@apoelstra would you mind reACKing?
